### PR TITLE
Unittest WebDavHelper

### PR DESCRIPTION
**Pull request by skshetry**
## Description

This adds a test for `WebDavHelper`, though we cannot check that directly. so, we check for the `lastRequest` made.

## Related Issue

- Part of #32527

## How Has This Been Tested?

- Locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

**Comment by skshetry** [comment]
The diff for changes from #33704 is here: [e7107fe..7d62aad](https://github.com/owncloud/core/compare/e7107fe..7d62aad)

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+			$this->client
+		);
+
+		$lastRequest = $this->history->getLastRequest();
+
+		$this->assertEquals(
+			'http://own.cloud/core/remote.php/dav/files/user1/folder/file%3Fq=hello%23newfile',
+			$lastRequest->getUrl()
+		);
+		$this->assertEquals(['http://own.cloud/core%3Fq="my files"%23L133'], $lastRequest->getHeaders()["Destination"]);
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: please make that line shorter

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	public function testMakeDavRequestOnBearerAuthorization() {
+		$response = WebDavHelper::makeDavRequest(
+			'http://own.cloud///core',
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: why extra multiple `/` here?

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	public function testGetDavPathForNewerDavPath() {
+		$davPath = WebDavHelper::getDavPath('user1', 2);
+		$this->assertEquals($davPath, 'remote.php/dav/files/user1/');
+
+		$davPath = WebDavHelper::getDavPath('user1', 2, 'files');
+		$this->assertEquals($davPath, 'remote.php/dav/files/user1/');
+
+		// $davPath = WebDavHelper::getDavPath(null, 2, 'files');
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: please delete this lines that do not run

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		// version 1 should be default
+		$davPath = WebDavHelper::getDavPath(null);
+		$this->assertEquals($davPath, 'remote.php/webdav/');
+	}
+
+	public function testGetDavPathForNewerDavPath() {
+		$davPath = WebDavHelper::getDavPath('user1', 2);
+		$this->assertEquals($davPath, 'remote.php/dav/files/user1/');
+
+		$davPath = WebDavHelper::getDavPath('user1', 2, 'files');
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: this and the test above is the same

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	public function withoutTrailingSlashUrlsProvider() {
+		return [
+			['http://own.cloud/', 'http://own.cloud'],
+			['http://own.cloud//index.php', 'http://own.cloud/index.php'],
+			['http://own.cloud//index.php//url', 'http://own.cloud/index.php/url'],
+			['http://own.cloud/login//login//', 'http://own.cloud/login/login'],
+			['http://own.cloud/login///login//', 'http://own.cloud/login/login'],
+
+			// get query should not have been sanitized
+			['http://own.cloud/login?redirect=//two.cloud//files', 'http://own.cloud/login?redirect=/two.cloud/files']
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: make the line shorter

File: tests/TestHelpers/Unit/WebDavHelperTest.php [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+		return [
+			['http://own.cloud/', 'http://own.cloud/'],
+			['http://own.cloud', 'http://own.cloud/'],
+			['http://own.cloud//index.php', 'http://own.cloud/index.php/'],
+			['http://own.cloud//index.php//url/', 'http://own.cloud/index.php/url/'],
+			['http://own.cloud/login//login//', 'http://own.cloud/login/login/'],
+			['http://own.cloud/login///login//', 'http://own.cloud/login/login/'],
+
+			// get query should not have been sanitized
+			['http://own.cloud/login?redirect=//two.cloud//files', 'http://own.cloud/login?redirect=/two.cloud/files/']
```
Comment (individual-it) [inline tests/TestHelpers/Unit/WebDavHelperTest.php]: shorter line pleae

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 2 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: fixed.

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 3 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: deleted. please review

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 4 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: Making sure that the defaults is set to `files`

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 5 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: done

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 6 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: done

**Comment by skshetry** [comment]
@individual-it please review again

File: tests/TestHelpers/Unit/WebDavHelperTest.php (hunk 1 above)
Comment (skshetry) [reply tests/TestHelpers/Unit/WebDavHelperTest.php]: done

File: tests/TestHelpers/HttpRequestHelper.php [hunk 7]
```diff
@@ -59,9 +59,13 @@ public static function sendRequest(
 		$config = null,
 		$cookies = null,
 		$stream = false,
-		$timeout = 0
+		$timeout = 0,
+		$client =  null
```
Comment (phil-davis) [inline tests/TestHelpers/HttpRequestHelper.php]: Needs PHPdoc of new parameter

File: tests/TestHelpers/DeleteHelper.php [hunk 8]
```diff
@@ -53,7 +53,8 @@ public static function delete(
 		$fileName,
 		$headers = [],
 		$davPathVersionToUse = 1,
-		$sourceIpAddress = null
+		$sourceIpAddress = null,
+		$client = null
```
Comment (phil-davis) [inline tests/TestHelpers/DeleteHelper.php]: Needs PHPdoc of new parameter

File: tests/TestHelpers/Unit/DeleteHelperTest.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+use TestHelpers\DeleteHelper;
+use TestHelpers\WebDavHelper;
+use GuzzleHttp\Client;
+use GuzzleHttp\Subscriber\Mock;
+use GuzzleHttp\Message\Response;
+use GuzzleHttp\Subscriber\History;
+
+class DeleteHelperTest extends PHPUnit_Framework_TestCase {
+	public function setUp() {
```
Comment (phil-davis) [inline tests/TestHelpers/Unit/DeleteHelperTest.php]: Add PHPdoc for each method. (I expect that the code style checker will complain)
It should be easy - mostly no parameters and returns nothing

File: tests/TestHelpers/WebDavHelper.php [hunk 10]
```diff
@@ -107,7 +107,8 @@ public static function makeDavRequest(
 		$sourceIpAddress = null,
 		$authType = "basic",
 		$stream = false,
-		$timeout = 0
+		$timeout = 0,
+		$client = null
```
Comment (phil-davis) [inline tests/TestHelpers/WebDavHelper.php]: PHPdoc needed for new parameter

**Review by phil-davis — COMMENTED** [review]
Needs various PHPdoc fixing

**Comment by phil-davis** [comment]
@skshetry needs rebase and resolve conflict

**Comment by skshetry** [comment]
@phil-davis @individual-it I have made fixes on docs and code style. Please review again.

**Review by phil-davis — DISMISSED** [review]
Seems OK.
@individual-it you had comments previously. Please check through them and approve if all is good.

**Review by phil-davis — APPROVED** [review]
Looks OK. We will see what CI says

**Review by individual-it — APPROVED** [review]

**Comment by skshetry** [comment]
Backport for `stable10` in #33876

**Pull request merged**